### PR TITLE
Model ArrayDecoding Builder as an enum

### PR DESCRIPTION
**Pull request by fdiaz**
This addresses this comment https://github.com/airbnb/ResilientDecoding/pull/11#discussion_r410524805

We're both not fans of the `Builder` approach, so this is an attempt to reword that with an `enum` instead to hopefully make things simpler without changing the API too much.

Let me know what you think @GeorgeLyon

### CODE REVIEW

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 1]
```diff
@@ -62,7 +62,7 @@ extension Resilient {
      */
     public subscript<T, U>(dynamicMember keyPath: KeyPath<ArrayErrorDigest<T>, U>) -> U
       where
-        Value == [T]
+      Value == [T]
```
Comment (fdiaz) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:65]: Indenting this file with Xcode

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 2]
```diff
@@ -147,11 +114,72 @@ struct ArrayDecodingError: Error {
         } else {
           return first.offset < second.offset
         }
-      }
-      .map { $0.element.element }
+    }
+    .map { $0.element.element }
```
Comment (fdiaz) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:118]: Indenting this file with Xcode

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 3]
```diff
@@ -147,11 +114,72 @@ struct ArrayDecodingError: Error {
...
+      self = .none(count: 1)
+    case .none(let count):
+      self = .none(count: count + 1)
+    case .arraydecodingError(let error, let count):
+      self = .arraydecodingError(error, count: count + 1)
+    }
+  }
+
+  /**
+   Attempts to build the `ArrayDecodingError`, returns `nil` if all elements decoded successfully.
```
Comment (GeorgeLyon) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:162]: I think it is telling that you used "Build" here, leading me to think `Builder` is actually the correct abstraction to use.

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 4]
```diff
@@ -147,11 +114,72 @@ struct ArrayDecodingError: Error {
...
   }
 
-  private var sortedErrorsAtOffset: [(offset: Int, error: Error)] = []
+  fileprivate var sortedErrorsAtOffset: [(offset: Int, error: Error)] = []
+}
+
+/**
+ Builds an `ArrayDecodingError`. This type assumes you are decoding elements in order, and care must be taken that the `mutating` members are called in the correct order.
+ */
+enum ArrayDecodingErrorState {
```
Comment (GeorgeLyon) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:127]: One thing I liked about `Builder` was that it was namespaced to `ArrayDecodingError` so we didn't have another top-level type. The other thing I liked was that the internal state of `Builder` was private, which is no longer the case if it becomes an `enum`.

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 5]
```diff
@@ -147,11 +114,72 @@ struct ArrayDecodingError: Error {
...
+    }
+    .map { $0.element.element }
   }
 
-  private var sortedErrorsAtOffset: [(offset: Int, error: Error)] = []
+  fileprivate var sortedErrorsAtOffset: [(offset: Int, error: Error)] = []
+}
+
+/**
+ Builds an `ArrayDecodingError`. This type assumes you are decoding elements in order, and care must be taken that the `mutating` members are called in the correct order.
```
Comment (GeorgeLyon) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:125]: "Build" again.

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 6]
```diff
@@ -147,11 +114,72 @@ struct ArrayDecodingError: Error {
...
+      self = .none(count: count + 1)
+    case .arraydecodingError(let error, let count):
+      self = .arraydecodingError(error, count: count + 1)
+    }
+  }
+
+  /**
+   Attempts to build the `ArrayDecodingError`, returns `nil` if all elements decoded successfully.
+   */
+  var error: ArrayDecodingError? {
```
Comment (GeorgeLyon) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:164]: Why does the `ArrayDecodingErrorState` have an `error`? Seems to me like the `ArrayDecodingErrorState` should be the state of an `ArrayDecodingError`, whatever that means.

File: Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift [hunk 7]
```diff
@@ -147,11 +114,72 @@ struct ArrayDecodingError: Error {
...
-  private var sortedErrorsAtOffset: [(offset: Int, error: Error)] = []
+  fileprivate var sortedErrorsAtOffset: [(offset: Int, error: Error)] = []
+}
+
+/**
+ Builds an `ArrayDecodingError`. This type assumes you are decoding elements in order, and care must be taken that the `mutating` members are called in the correct order.
+ */
+enum ArrayDecodingErrorState {
+  case start
+  case none(count: Int)
```
Comment (GeorgeLyon) [inline Sources/ResilientDecoding/ResilientPropertyLevelErrors.swift:129]: I'm not sure the difference between `start` and `none` is immediately apparent, which is exacerbated by the fact that the internal state of this thing is no longer `private`.

**Review by GeorgeLyon — CHANGES_REQUESTED** [review]
Thanks for putting this together, but I'm not sure this simplifies things. Most of my feedback is qualitative, but I think leaking the internal state of `Builder/State` and not having it namespaced are strict negatives.
Even if I weren't leaning negative on this, I think we should have a higher bar for no-behavior-change PRs (to clarify, I consider the removal of a public underscored protocol and conformance from my previous PR to be "behavior change") and in my mind that high bar should be fixing some concrete issue for folks contributing to the repository, which I'm not sure this PR meets. Though I have a feeling that the last statement will come back to bite me a few PRs from now...

**Comment by fdiaz** [comment]
> I think leaking the internal state of Builder/State [...] strict negatives

I thought this was what we wanted instead of having a mutable index internally, but I think I misunderstood.

 > and not having it namespaced are strict negatives.

This could have been easily fixed by putting it back inside the namespace. There's nothing technically preventing us from doing this. I usually think of things "creating other types" to be "higher level" than those types and not internal to that type which is why I moved this outside.

> Even if I weren't leaning negative on this, I think we should have a higher bar for no-behavior-change PRs

I'm not sure I follow. This doesn't do any behavior changes in the public facing API.

> and in my mind that high bar should be fixing some concrete issue for folks contributing to the repository

I was hoping this type was easier to maintain given we both didn't didn't like the `Builder`

Since your reaction is clearly negative to this path I'm gonna put this idea off.

I'd love to get rid of the Builder at some point since that API is not great, but since its internal it's not a big problem either.

**Pull request closed without merging**